### PR TITLE
Add Access Level-Based Contextual Retrieval Augmented Generation

### PR DESCRIPTION
**Pull request by arkanalexei**
This PR finally creates the RAG system and also incorporating access-level-based contextual retrieval for chat responses. Key updates include:

- Modifications to SlackAdapter to retrieve user access levels and pass them with user queries.
- Refactoring of ChatEngine and related classes (ChatOpenAI, ChatAnthropic) to support context retrieval based on access levels.
- Enhancements to ChatEngineSelector and Retriever for querying contextual data from PostgreSQL based on access permissions.

### CODE REVIEW

**Comment by nyoosteven** [comment]
Okay, I already checked Uttsada code. It seems correct.

**Comment by nyoosteven** [comment]
@arkanalexei how do you use level based RAG. can you give me more information regarding this? Thanks!

**Comment by arkanalexei** [comment]
> @arkanalexei how do you use level based RAG. can you give me more information regarding this? Thanks!

To put it simply, each user that uses the `/askbot` command will have their slack information parsed and their access level will be passed into the `query` function. If a user is not registered in broom's database then they will have the lowest `access_level` = 1.

**Comment by jasonirvine76** [comment]
LGTM

File: rag/retriever/retriever.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+            input=query,
+            model="text-embedding-3-small"
+        )
+        query_vector = embedding_result.data[0].embedding
+        
+        return self.postgres_handler.query(query_vector, access_level=access_level, top_k=top_k)
+
+    def _retrieve_context_tabular(self, query, access_level) -> str:
+        del query, access_level
+        return ''
```
Comment (Insta-x) [inline rag/retriever/retriever.py:19]: I'm curious, what's the point of this code? why not simply return empty if it's still not implemented for now?

**Comment by Insta-x** [comment]
Looks good to me, no weird issue or code. Just need to fix the conflicting files.

**Pull request merged**